src/lib.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2017 King's College London
+// created by the Software Development Team <http://soft-dev.org/>
+//
+// The Universal Permissive License (UPL), Version 1.0
+//
+// Subject to the condition set forth below, permission is hereby granted to any person obtaining a
+// copy of this software, associated documentation and/or data (collectively the "Software"), free
+// of charge and under any and all copyright rights in the Software, and any and all patent rights
+// owned or freely licensable by each licensor hereunder covering either (i) the unmodified
+// Software as contributed to or provided by such licensor, or (ii) the Larger Works (as defined
+// below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software is contributed
+// by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create derivative works
+// of, display, perform, and distribute the Software and make, use, sell, offer for sale, import,
+// export, have made, and have sold the Software and the Larger Work(s), and to sublicense the
+// foregoing rights on either these or other terms.
+//
+// This license is subject to the following condition: The above copyright notice and either this
+// complete permission notice or at a minimum a reference to the UPL must be included in all copies
+// or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+use std::collections::hash_map::{Entry, HashMap};
+use std::hash::Hash;
+use std::sync::Mutex;
+
+const HOT_THRESHOLD: u16 = 10;
+
+/// The current meta-tracing state of a given location in the end-user's code.
+#[derive(Debug, PartialEq)]
+enum LocPhase {
+    /// Below `HOT_THRESHOLD`, but counting up towards it.
+    Counting(u16),
+    /// A trace is being collected.
+    Tracing,
```

Comment:
Something doesn't sit right with me about `Tracing`/`Trace`. These are pretty similarly named.

How about `BeingTraced` and `TraceReady`? (and eventually `TraceCompiling`? see comments below about this state)

Reply:
How about: Counting, Tracing, Compiled?

In my head, at the moment, I'm including everything about compilation under "tracing". Put another way, "tracing" means "don't start another trace". Maybe we will need another state, but perhaps it's better to wait until we're sure?

Reply:
The names are perfect, let's use those.

I think without the distinction you are likely to spawn duplicate compilation threads. It works now because traces are assumed to appear synchronously.

It's up to you whether you introduce the new state now or in the PR that introduces compilation threads.